pkg/gpu/nvidia/mig/mig.go

```diff
@@ -0,0 +1,357 @@
+	}
+
+	gpuFileRegexp := regexp.MustCompile("gpu([0-9]+)")
+	giFileRegexp := regexp.MustCompile("gi([0-9]+)")
+	deviceRegexp := regexp.MustCompile("DeviceFileMinor: ([0-9]+)")
+
+	for _, capFile := range capFiles {
+		m := gpuFileRegexp.FindStringSubmatch(capFile.Name())
+		if len(m) != 2 {
+			// Not a gpu, continue to next file
+			continue
+		}
+
+		gpuID := m[1]
+
+		giBasePath := path.Join(nvidiaCapDir, capFile.Name(), "mig")
+		giFiles, err := ioutil.ReadDir(giBasePath)
+		if err != nil {
+			return fmt.Errorf("unable to discover gpu instance: %v", err)
+		}
+
+		for _, giFile := range giFiles {
+			if !giFileRegexp.MatchString(giFile.Name()) {
+				continue
+			}
+
+			gpuInstanceID := "nvidia" + gpuID + "/" + giFile.Name()
+			giAccessFile := path.Join(giBasePath, giFile.Name(), "access")
+			content, err := ioutil.ReadFile(giAccessFile)
+			if err != nil {
+				return fmt.Errorf("unable to read GPU Instance access file (%s): %v", giAccessFile, err)
+			}
+
+			m := deviceRegexp.FindStringSubmatch(string(content))
+			if len(m) != 2 {
+				return fmt.Errorf("unexpected contents in GPU instance access file(%s): %v", giAccessFile, err)
+			}
+			giMinorDevice, _ := strconv.Atoi(m[1])
+
+			ciAccessFile := path.Join(giBasePath, giFile.Name(), "ci0", "access")
+			content, err = ioutil.ReadFile(ciAccessFile)
+			if err != nil {
+				return fmt.Errorf("unable to read Compute Instance access file (%s): %v", ciAccessFile, err)
+			}
+
+			m = deviceRegexp.FindStringSubmatch(string(content))
+			if len(m) != 2 {
+				return fmt.Errorf("unexpected contents in compute instance access file(%s): %v", ciAccessFile, err)
+			}
+			ciMinorDevice, _ := strconv.Atoi(m[1])
+
+			gpuDevice := path.Join(d.devDirectory, "nvidia"+gpuID)
+			if _, err := os.Stat(gpuDevice); err != nil {
+				return fmt.Errorf("GPU device (%s) not fount: %v", gpuDevice, err)
+			}
+
+			giDevice := path.Join(d.devDirectory, "nvidia-caps", "nvidia-cap"+strconv.Itoa(giMinorDevice))
+			createDeviceNode(giDevice, uint32(majorDevice), uint32(giMinorDevice))
+			if _, err := os.Stat(giDevice); err != nil {
+				return fmt.Errorf("GPU instance device (%s) not fount: %v", giDevice, err)
+			}
+
+			ciDevice := path.Join(d.devDirectory, "nvidia-caps", "nvidia-cap"+strconv.Itoa(ciMinorDevice))
+			createDeviceNode(ciDevice, uint32(majorDevice), uint32(ciMinorDevice))
+			if _, err := os.Stat(ciDevice); err != nil {
+				return fmt.Errorf("Compute instance device (%s) not fount: %v", ciDevice, err)
+			}
+
+			d.gpuPartitions[gpuInstanceID] = []pluginapi.DeviceSpec{
+				{
+					ContainerPath: gpuDevice,
+					HostPath:      gpuDevice,
+					Permissions:   "mrw",
+				},
+				{
+					ContainerPath: giDevice,
+					HostPath:      giDevice,
+					Permissions:   "mrw",
+				},
+				{
+					ContainerPath: ciDevice,
+					HostPath:      ciDevice,
+					Permissions:   "mrw",
+				},
+			}
+
+		}
+	}
+	return nil
+}
+
+func createDeviceNode(path string, major, minor uint32) error {
```

Comment:
nit: the function name `createDeviceNode()` is a bit confusing because `CreateDeviceNodes()` seems to be doing way more work than this one does.